text/0000-dynamic-import-hints.md

```diff
@@ -0,0 +1,264 @@
+---
+title: Dynamic Import Hints and Its Metadata
+status: DRAFTED
+created_at: 2020-03-16
+updated_at: 2020-03-22
+rfc: https://github.com/salesforce/lwc-rfcs/pull/27
+champion: Aliaksandr Papko (@apapko) | LWC Team
+implementation: https://git.soma.salesforce.com/lwc/lwc-api/pull/42 & https://git.soma.salesforce.com/lwc/lwc-platform/pull/361
+---
+
+# Hints Syntax and Its Metadata
+
+## Summary
+This RFC suggests authoring syntax for specifying dynamic import hints and describes its metadata shape.
+
+## Motivation
+Using hints allow component authors to declare the component preload context by parameterizing dynamic imports. The values from the hints can be used to optimize dependency resolution and pre-fetching.
+
+## Hint Invariants
+* Hint key and value are string literals
+* Hint keys and values must be statically verifiable i.e. no curly braces to implicitly refer to a variable and such
+* Multiple hints will be allowed but with restrictions
+
+## Platform Specific Hint Invariants
+For the salesforce platform, the following invariants are proposed
+* Hint values must be globally available and resolvable on the server side
+* Hint values must be finite and enumerable (ex: no Math.random())
+
+## Detailed Design
+
+### Grammar
+There will be grammar restriction in place for hints so that the compiler can parse the LWC modules, gather the metadata and provide that as a result of the compilation.
+
+#### Hint Syntax
+A hint is expressed as a comment inside a dynamic import expression. The hint syntax is either a line comment or a block comment with a *key* as a resource and a *value* as a resolved resource value. The key and value will be separated by a colon(`:`) and delimited using double quotes(`"`).
+
+```js
+/* "<key>": "<value>" */
+```
+Hint comment e.g.:
+```js
+/* "@salesforce/client/formFactor": "SMALL" */
+/* "MY_CONDITION": "VALID" */
+// "MY_OTHER_CONDITION": "NOT_VALID"
+```
+
+Dynamic import with a hint e.g.:
+```js
+export async function example1() {
+    loadedModule = await import("lightning-foo"/* "@salesforce/client/formFactor": "SMALL" */);
```

Comment:
Would this example be better if it showed an `if` statement wrapping the import as well as the import of `@salesforce/client/formFactor`? To make it clear that the component author is variably importing something and the app framework will react to that?